src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs

```diff
@@ -398,4 +415,99 @@ public static bool TryGetHandshakeProtocol(SslProtocols protocols, [NotNullWhen(
         version = null;
         return false;
     }
+
+    public static void AddConnectionEndReason(IConnectionMetricsTagsFeature? feature, ConnectionEndReason reason)
+    {
+        Debug.Assert(reason != ConnectionEndReason.Unset);
+
+        if (feature != null)
+        {
+            if (TryGetErrorType(reason, out var errorTypeValue))
+            {
+                feature.TryAddTag(ErrorType, errorTypeValue);
+            }
+        }
+    }
+
+    public static void AddConnectionEndReason(ConnectionMetricsContext? context, ConnectionEndReason reason, bool overwrite = false)
+    {
+        Debug.Assert(reason != ConnectionEndReason.Unset);
+
+        if (context != null)
+        {
+            if (TryGetErrorType(reason, out _))
+            {
+                if (context.ConnectionEndReason == null || overwrite)
```

Comment:
We only overwrite if there's an error type?  That is, we won't clear the error type if the new end reason indicates success and `overwrite` is true?

Reply:
overwrite would only be true when the reason is an error.

Anyway, I changed this code to make it more understandable.